### ISSUE TITLE
Repeated call duration (Answer: Microsoft has introduced massive database bugs in the logging of calls, would take days of hard work to analyze it all and devise a workaround, volunteers and pull requests are welcome, read this entire chat log)

### ISSUE THREAD

**Issue by hectorpal** [Temptin/SkypeExport#15]
I'm obtaining repeated duration in a ChatHistory. I detected in for a particular users. It's not from the beginning but for calls from Oct 1st 2015. It may have coincided with the relase of Skype 7.13 for Mac
on September 30th, 2015, that added support for OS X El Capitan (10.11). I'm not including the main.db because of privacy. I'm looking at the code, and hasn't been able to ping point it. I'm under OS X. I tried both with the distributed binaries and one compiled on my own. The html generated by both of them was identical, except time stamp of generation. Valgrind reported not error.

**Comment by hectorpal** [comment]
I understand you the package may not be under maintenance. I'm ready to offer a patch if I figure out this.

**Comment by hectorpal** [comment]
1930 is the repeated duration. This query is only returning one row:
`SELECT * FROM Calls WHERE duration = 1930 ORDER BY begin_timestamp DESC`

It seems the duration was 'remembered', but I see in the code it is reseted to 0 each time.

**Comment by hectorpal** [comment]
I noticed that in Calls, for that particular skypeid, the field 'conv_dbid' is zero since 2015-10-02 18:38:55, that coincides with my observation.

`SELECT  duration, conv_dbid, begin_timestamp, host_identity, current_video_audience FROM Calls WHERE current_video_audience = 'such-skypeid ' ORDER BY begin_timestamp DESC`

It seems I have a good lead: what should the code if conv_dbid is zero?

**Comment by hectorpal** [comment]
I discover far more inconsistencies. Sometimes, when what Skype report as 'unanswered', the exported htm says it was a call with the duration of the previous call, and that over some more unanswered calls. Let's say right now it's not exactly useful for legal porpoises.

**Comment by Temptin** [comment]
Hi, you're right that the project isn't maintained by me anymore. I barely ever log in here anymore either, I just happened to see this now. I've given Santosh write-access so that he can implement his GUI patch if he ever has time for that, and he can merge pull requests.

I can see the reason for the behavior. You say that Skype has stopped logging the conv_dbid... I hate when they do things like that. They sneak in database-corrupting changes like that more and more frequently after Microsoft's acquisition. I spent two months last year fixing all of Microsoft's database corruptions, and you've just found another thing they messed up.

Here's why you are getting the 1930 duration over and over again:
`SELECT duration FROM Calls WHERE (conv_dbid=? AND begin_timestamp<=?) ORDER BY begin_timestamp DESC LIMIT 1`

It finds the latest Calls table event that matches the conversation ID and is **older than** or equal to the call event timestamp in the chat log itself.

Since Microsoft has messed up the Skype code yet again and have stopped properly writing the conversation ID to the Calls table, all newer calls will be seeing the last call which had a proper conv_dbid value.

The Calls table has a "call_guid", a globally unique ID for every call... So why did I ignore that and write my own method that was perfectly reliable until Microsoft messed it up by no longer logging the conversation ID? Well, because... this event "type:39" code comment explains why I had to do it that way:

```
* there is a column, call_guid, which identifies each call for Skype5
but it's NULL for VERY OLD calls performed in Skype versions older
than 5, so we cannot rely on it under any circumstance!
```

So... since Microsoft has now decided to only log/use the call_guid, we're in a dilemma: Old databases don't have it, and new databases only have it...

The solution would be something like this:
`SELECT duration FROM Calls WHERE (call_guid=? OR (conv_dbid=? AND begin_timestamp<=?)) ORDER BY begin_timestamp DESC LIMIT 1`

A GUID in the Skype database is a sequence of 32 unsigned bytes. Except in the case of calls (I just checked) it's not that; it's actually a text string, such as: "141206", or "50742" or even "549e712c-ed83-4a5e-8ea1-a4e5620bbf9e" or "60c96eb93a8181550a3da2e72f598595" (lovely consistency, Microsoft! Great job!).

The code would need plenty of changes... in fact I started to write a list of them here but they're just so many that it takes me longer to find all the code locations and describe the changes necessary there, than to just do it myself...

**Comment by Temptin** [comment]
I definitely see what you mean, in my own Calls table I see conv_dbid in old calls but not in the latest ones.

**Comment by hectorpal** [comment]
Thank you for having a look on this. Quick question, does Skype keeps the local databases updated, even if some calls by the same user were initiated in another device? For example, the database is almost the same if you have skype in a laptop and a tablet.

**Comment by Temptin** [comment]
Yeah they do that via cloud syncing, and it's very unreliable. Whenever your desktop Skype connects, it grabs all events via Microsoft's servers, and that's the leading cause of corruption (losing fields and events), since their cloud history isn't the full event (it's like a small subset). The highest-quality data in the database is whatever happened on your actual computer. The synced data is spotty and has holes. That's part of the reason I had to work for 2 months rewriting a lot of things to handle holes in the sync data (like vital fields suddenly being NULL for non-local events).

In other news... the rewrite seems to be done and should be written perfectly, but I have not compiled or tested it. I will upload the commit in a moment and you can compile and check.

**Comment by hectorpal** [comment]
Good :). I'll check it later.

**Comment by Temptin** [comment]
It took me time to set up git for writing to the GitHub repo again. It's been a while.

The commit is up! :)

https://github.com/Temptin/SkypeExport/commit/60f1339688f647159eb1ff18981e9866cea65e50

Please let me know how it goes. It should now be able to find Microsoft calls as well as all old-style call logs.

Thanks for reporting this with such great detail. Knowing that your conv_dbid was zero was the key to realizing what had happened. I think this change also fixes issue #12, where the user was seeing "Call ended, no answer". That's what we output when the duration is 0. And since we set duration to 0 and then retrieve the real duration from the DB, and his query failed to find the call (due to Microsoft's database corruption), it _would_ indeed think the duration was 0 (aka "No answer"). So this change should fix that ticket too.

After you've checked the changes to see if they work as expected, then I'll be able to close two bugs...

Unfortunately I will not be making any precompiled versions for Windows or Mac. I could technically set up a Mac compiler environment again, but it doesn't seem worth it to just update one OS version since I have no access to compiling for Windows anymore. :/

If you or someone else can volunteer to compile for Windows, I would be happy to build the new Mac version. I'll be making a new tagged version for 1.5.0 now...

**Comment by Temptin** [comment]
Okay, v1.5.0 is committed, with the change. Check it out at your own leisure, and remember that there's no new pre-built version, just the commits:

https://github.com/Temptin/SkypeExport/commit/398b45c78a0290fe9ffaf1c029bf0f5737fddb0e

If it works as it should, then that'll be the next 1.5.0. Can you build a Windows 32-bit binary? If so, I'll make the Mac build. It'd be nice to get this fix out to people to show proper call durations again, but I'm not installing a whole VM application and Windows and Visual Studio (hours of my life) just for this. ;)

**Comment by hectorpal** [comment]
I'm testing in OS X, and compiled boot myself.

In my test, the new generated htm is very similar to the previous one, including repetition of dates since Oct 1st 2015. Just 4 of 'callicons' lines changed of 684. The guilty lines are around 170. This means the commit didn't improved much for this test.

Also, I've mentioned another problem that I think is unrelated. In the beginning of my file, I'm obtaining also repeated durations for dates around Apr 2015. But those corresponds what Skype reports as unanswered-calls. For example, Skype says I talk during 40m, then later I made two unanswered calls, and then I made a successful call of 20m. The program is now reporting four calls, the first three of 40m, reporting 'Call ended' in those three calls.

(Btw, I don't have an developing environment in Windows)

**Comment by Temptin** [comment]
That's weird... I didn't totally blindly test this, I used sqlite3 on the command line and did "SELECT call_guid FROM Messages" to get one of the call GUIDs from my own database, and then I tried the new SQL statement and was able to find the call that way (via GUID). But maybe it's not working in practice... Can you investigate this further? I have no time or interest in programming anymore and quit last summer. :S

**Comment by Temptin** [comment]
What to investigate:
- Look at my commit (https://github.com/Temptin/SkypeExport/commit/60f1339688f647159eb1ff18981e9866cea65e50), and you will see that we now get "call_guid" (a char\* string pointer or null pointer, given to us by Sqlite) for every message event we parse.
- We then use that call_guid if non-null, to get the call info, by looking in the 'name' column of the Calls table. It's a weird column name, but that's the only Calls table column that contains the "call_guid" (Messages table) string for me in my own database. It doesn't seem to be stored anywhere else...

And I tested this in sqlite3 manually and got the correct call... So... with this change, it really should be fixed... unless it's somehow not properly copying the GUID string during the parameter binding here: https://github.com/Temptin/SkypeExport/commit/60f1339688f647159eb1ff18981e9866cea65e50#diff-8f74996125abdb7c038604bb0975170cR919

I won't check this any further since I already spent 2 hours and have uninstalled the dev environment again, but if you find the cause, you'll be a hero and I'll merge it for you. ;)

**Comment by Temptin** [comment]
I think I know what's wrong... Skype doesn't add Calls table entries for every call (perhaps missing for the ones synced from other devices), or something like that...

Look at your own database in a SQLITE viewer, like sqlite3 on the command line (on main.db). Then try these queries:

Get all call GUIDs (this lists all call start/call end events, and gets all unique values found in the call_guid column).
`SELECT DISTINCT call_guid FROM Messages WHERE call_guid not null AND (type=30 OR type=39);`

Next, pick one of them and use this query (putting your own call_guid value in the name parameter):
`SELECT * FROM Calls WHERE (name not null AND name != '' AND name='a92e0086-a76e-4a30-938c-edbe748256c5');`

Note that in my case, the first query gives me all of these call IDs:
549e712c-ed83-4a5e-8ea1-a4e5620bbf9e
54d7946f-9861-4845-9eab-d78d3174c8c7
60c96eb93a8181550a3da2e72f598595
950a7d075bc3222baedee8662a6cca74
a92e0086-a76e-4a30-938c-edbe748256c5
a9ca552e09788448b29a3e1f320fa774

But only a few of them exist in the Calls table. In fact, let me show you what's in there:
`SELECT DISTINCT name from Calls;`
1-1398238199
1-1431876883
54d7946f-9861-4845-9eab-d78d3174c8c7
a92e0086-a76e-4a30-938c-edbe748256c5

So... it seems this is the issue:
- Microsoft has screwed up Skype by no longer logging conv_dbid in the Calls table.
- They have also screwed it up by not logging every call in the Calls table.

There's nothing we can do if there's no data in the database. ;)

**Although, something just hit me: Maybe the entry is missing in the Calls table if there was 'No answer'. That could hopefully explain why my own Calls table only has two of the 6 calls. In that case everything will work CORRECTLY and use a duration of 0 ('No answer' will be output by SkypeExport for those calls that are not in the Calls table). So no, I don't think we need to worry that only 2 of my calls are in my Calls table; it's probably just that nobody answered those times. I have done a lot of call attempts with no answer.**

**Comment by Temptin** [comment]
Ok I think I have figured it out already...
- Skype might only log a Calls table entry if the call was picked up.
- For all other calls it creates a call_guid, but nothing is put in the Calls table.
- The scanner tries to find the call_guid in the 'name' column of the Calls table, but fails...
- So it falls back to the "conv_dbid method, and finds older calls for that conversation where you still had a conv_dbid.

So now please investigate before I change a single line of code again: Please check if the 'wrong duration' calls for you are ones that weren't even picked up. In that case, I will delete the legacy-code, so that we only use call_guid from now on. I don't think many people care about getting the call durations for Skype 1-4 database events. That's ages ago. Those calls will be exported as "No answer" if I decide to remove the legacy code, since it seems pretty clear that no call is logged if the user didn't answer.

Here's how you can do this check:
- Check in Skype's official client and see if IT says "No answer" for your calls with the wrong duration.
- Look in your Messages table for that particular call event and see what its call_guid is. A good way to find it is to find the message with a "body_xml" value that contains some text from a line of text before/after the call, then look at surrounding "type=30" or "type=39" events in the Messages table, and look in their call_guid column
- Check if your Calls table has that call_guid in its 'name' column anywhere.

I am strongly suspecting no call is logged if nobody picked up.

**Comment by hectorpal** [comment]
Calls is very sparsed populated. All this is for studying calls with one user, and it doesn't appear there. I did `SELECT call_guid, timestamp FROM Messages ORDER BY timestamp` and saw that old calls tend to not have call_guid. I couldn't determine the limit, but later they are more popular.

**Comment by hectorpal** [comment]
I'm checking what you asked right now...

**Comment by Temptin** [comment]
@hectorpal: Yes Calls is very sparse for me too, only 4 entries.

But your method for checking call_guid in Messages is wrong. It looks at non-calls too, that is why you get lots of blank lines. Use this statement I posted above:
```SELECT DISTINCT call_guid FROM Messages WHERE call_guid not null AND (type=30 OR type=39);```

And this to look at which IDs are in your actual Calls table:
```SELECT DISTINCT name from Calls;```

**Comment by Temptin** [comment]
@hectorpal: Yes Calls is very sparse for me too, only 4 entries. Compared to having lots more call_guids in the Messages table...

But your method for checking call_guid in Messages is wrong. It looks at non-calls too, that is why you get lots of blank lines. Use this statement I posted above to get all call events with non-null value:
`SELECT DISTINCT call_guid FROM Messages WHERE call_guid not null AND (type=30 OR type=39);`

And use this one (just for safety check) to get all call events that have no call_guid value:
`SELECT call_guid FROM Messages WHERE call_guid is null AND (type=30 OR type=39);`
(For me this one gives 0 results, as it should. Every call start/end type event has a call_guid value.)

And this to look at which IDs are in your actual Calls table:
`SELECT DISTINCT name from Calls;`

**Comment by Temptin** [comment]
My guess is that I have to drop Skype 1-4 call duration support (those will all say "No answer"), since every Skype from 5 and higher logs a call_guid value in the Messages table. And I cannot have the old support without accidentally making it all revert to the old-school scanner whenever a call_guid isn't in the Calls table (i.e. when nobody picked up... if that is indeed when it doesn't log anything). Luckily Skype 4 is from something like 2010, and only people with extremely long histories would ever be trying to export such old calls. They'll now see "no answer" for 6+ year old events, but it's the only proper fix if we're going to nowadays rely on call_guids instead. We can't fall back to "close enough" anymore, or we get the false matches you just saw...

**Comment by hectorpal** [comment]
1. On Apr 12 2015, I have repeated durations in the SkypeReport but "No Answer" in Skype official.
2. The same date, I'm obtaining the correct calls and durations from

> SELECT  duration, conv_dbid, begin_timestamp, host_identity, current_video_audience FROM Calls WHERE current_video_audience = 'ThatUserName' ORDER BY begin_timestamp DESC

and it doesnt't include the non-answered calls.

I'm checking the rest, following your most recent comment.

**Comment by Temptin** [comment]
Ok, repeated durations means:
1. It's not finding the call_guid in the Calls table's 'name' column.
2. It's then trying to find a Call with a conv_dbid matching the conversation and a timestamp <= the call_end event of the Messages table.
3. It finds an old call that way, that's why you see the repeated durations.

When you see "No Answer" in the official Skype, it means a duration of 0. In the past, that was always logged in the Calls table too. It seems they changed it to only log picked-up calls. Thanks for confirming that.

The fix will be that I remove Skype 1-4 support, and instead use ONLY 'name' (the call_guid) value. Please check this while I work on the rewrite:

What's your call_guid in Messages for the ANSWERED calls in that conversation?
Is that call_guid found in the 'name' column of your Calls table? It should be.

**Comment by hectorpal** [comment]
What about an option for supporting Skype 1-4? You can warn the user that reports up to some date should be ok, or reports before some date should be ok. Desperate people could run both version and concatenate the results.

I located the timestamp of a correct _answered_ call of Apr 12th 2015 in Calls. 

> SELECT duration, conv_dbid, begin_timestamp, host_identity, current_video_audience FROM Calls WHERE current_video_audience = 'ThatUserName' ORDER BY begin_timestamp DESC

Then I used that `being_timestamp`, and ran

> SELECT  call_guid FROM  Messages where timestamp = found-timestamp

I got a  call_guid.

Then I ran in Python

> SELECT \* FROM Calls

Printed the records, and grep the output, looking for such call_guid. Nothing. So, if I did it right, the answer is _no_.

(I hope utf didn't got in the middle)

**Comment by hectorpal** [comment]
(@Temptin, I updated the previous comment)

**Comment by Temptin** [comment]
You did that wrong; the timestamp is always different between Calls table and Messages table :) That's why I used "begin_timestamp <=" in the past scanning method. And that is why you didn't find anything.

Look at 'name' in the Calls table. That's the same as 'call_guid' in the Messages table. Match those two to each other. Don't use video_audience or conv_dbid or timestamps or anything else. :)

And no, I can't support Skype 1-4 anymore. Microsoft has done this: Rely entirely on call_guid, Delete conv_dbid, and no longer log anything at all into Calls if there's no call.

The old scanning method relied on the fact that every call was logged and had perfect conv_dbid information. Now it's no longer the case. It's broken Skype 1-4 permanently. Trying to support it breaks Skype 5+ support (as you saw). But that's no big deal; calls 6 years ago just show "No duration" now. I can't do anything else. Not even official Skype has support for those old calls anymore, precisely because they changed so much about how calls are logged.

**Comment by Temptin** [comment]
I think someone at Microsoft thought it was wasteful to log a Calls table entry for every non-answered call. Thanks to that person we no longer support Skype 1-4 call durations. Trying to still support it totally breaks Skype 5+ support (as you saw). It's a lot of factors in their changes that make backwards-compatibility near-impossible to solve. It's such a minor thing though, I'm not losing sleep over it, the world moves on from things that happened 6+ years ago. I am finishing up the new version that only supports Skype 5+ call durations and uses the permanent, modern method of matching by GUID instead.

**Comment by hectorpal** [comment]
Trying again, I ran
> SELECT duration, conv_dbid, begin_timestamp, name FROM Calls WHERE current_video_audience = 'ThatUserName' ORDER BY begin_timestamp DESC

and located appropriate 'name', corresponding to the answered call.

Such name *is* in Calls.

**Comment by Temptin** [comment]
Thanks a lot, perfect. Confirms my own observation that they no longer log Calls at all if there is no answer. And by the way, call_guid+name combo has been perfectly stable since Skype 5. I was the only one that supported Skype 1-4 call durations. The official Skype client itself didn't support them and didn't even show them. :) SkypeExport does so much more than the official Skype, and this is the first time I had to drop some very old legacy support for a very minor feature. Sad day but it's unavoidable this time.

I am nearly done with the rewrite now. I'm going to force-push a totally rewritten commit log because of the problem with the earlier fix (wouldn't have written it that way if I knew Microsoft no longer logs non-answered calls, so it's unfortunate that I have to change the history). Oh well nobody has pulled this project in the meantime, except you. :)

One moment...

**Comment by hectorpal** [comment]
Yes, supporting Skype+ calls should be ok. At least, it would be clear.

Hum, I'm obtaining a 'name' from Calls with the right duration, but when dumping 'call_guid', I cannot find such 'name'.  @Temptin , are you sure this is ok? You may have read a comment I deleted later.

**Comment by Temptin** [comment]
> Hum, I'm obtaining a 'name' from Calls with the right duration, but when dumping 'call_guid', I cannot find such 'name'.

That's database corruption. Common with Skype after Microsoft bought it and added their messy, buggy "Cloud history" syncing feature.

**Comment by Temptin** [comment]
Microsoft themselves use the call_guid + name columns. Skype has been like that since version 5. All other Calls table columns are merely inaccurate timestamps (not matching Messages table events) + the conversation ID that the call came from. Microsoft no longer even logs the conversation ID. So the only way to find a call is via call_guid == name. No other columns uniquely identify a call. And as you saw, the older scanning method I used became a mess due to Microsoft not logging unanswered calls and not logging convo_dbid. Annoying... yes.

**Comment by Temptin** [comment]
Sorry for the mess earlier. Now I have pushed the rewrite which only uses the GUID:

https://github.com/Temptin/SkypeExport/commit/6016f7e189d58215c546d6ee498db91db183e27b

Please try exporting via that now.

**Comment by hectorpal** [comment]
Yes, annoying.

Interesting enough, I'm obtaining just 378 entries from

> SELECT duration FROM  Messages, Calls where Messages.call_guid = Calls.name  ORDER BY duration;

but only 230 is duration is not null.

But it seems the durations appears twice. 

> SELECT distinct duration FROM  Messages, Calls where  Messages.call_guid = Calls.name and duration is not null ORDER BY duration;

Reports only 107.

**Comment by hectorpal** [comment]
Trying with the new code.

**Comment by Temptin** [comment]
It's a big mess... what has Microsoft done. Some geniuses working there... This is why I had to spend 2 months of last year creating workarounds to catch up with all the database bugs they introduced ;)

**Comment by hectorpal** [comment]
Now I'm obtaining just a few durations. Under 20. Copy/Paste from Skype official is around 245. :(

**Comment by Temptin** [comment]
FOOOOOOOR FUCK'S SAKE.............. Sorry for swearing but this has been 3 hours of my life on an 'easy' fix and still no progress... Microsoft has screwed this up badly.

This is how Skype always worked:
- conv_dbid = Must exist, is the exact conversation ID the call was made in. Microsoft = deleted it, it is now always 0/null.
- Calls = always logged in the Calls table, even if not answered. Microsoft = Most calls are not logged. The calls that are not logged show up as "No answer" in the official Skype client, leading me to believe it no longer logs unanswered calls, to save database space.
- name = This was added in Skype 5 as a perfect way for Skype to map calls to their Messages table events. So call_guid in the Messages table == name in the Calls table. This was the proper way to find calls. Skype itself couldn't find Skype 1-4 call details in its own database logs, due to this change. But I wrote SkypeExport with great care to add backwards compatibility...

But Muckroshits changes constantly break Skype's database, and this is no exception. The changes are too big for the old scan method to work or even be rewritable. And now something is wrong with how they map call_guids...

Shit... can you look at your database and try to figure out how else they may be mapping calls? Because in my own databsae I have only a handful of calls, and the GUIDs are text strings in the format 'a92e0086-a76e-4a30-938c-edbe748256c5', and there are about 8 strings in my Messages table, and only 2 rows (total!) in my Calls table, which proves they no longer log anything...

And if I look at my own database, the only Calls table column that contains the GUID is the 'name' column. And that is exactly how it worked back in Skype 5 too... So that hasn't changed... But how the hell are Microsoft finding the calls then?

What is going on here... what have I gotten myself involved in. This was gonna be a quick 10 line fix and now I'm annoyed as hell. :) Not at you of course. AT MICROSOFT!

**Comment by Temptin** [comment]
The initial issue was that we no longer found calls since they didn't have conv_dbid anymore in the Calls table... That was 3+ hours ago... So the old scanner doesn't work AT ALL for Microsoft's new calls. So I added the call_guid method instead, but since those calls weren't found either, it reverted to the old code, and you still saw repeated durations (plus a few new durations that couldn't be found at all earlier due to Microshit's lack of conv_dbid values). And so I moved entirely to call_guid, which would have worked in Skype 5... and now we find almost nothing. What has Microsoft done!?

Microsoft.
Stop.
Messing.
Up.
Skype... argh.

**Comment by Temptin** [comment]
By the way... I just looked at Skype on my phone vs on my desktop.

My phone shows "Call, no answer" for a call attempt made via the phone. My desktop Skype shows NOTHING. It doesn't even Cloud History sync those unanswered calls. :)

Here's the problem: **Microsoft's engineers bought Skype. They had no idea how the codebase worked. And they have continually fucked it up and hacked it like this. This was the worst one yet...**

**Comment by hectorpal** [comment]
So, they are not even in Sync.

Ok, I couldn't find the appropriate XKCD where a guy start working in a program so he avoid some tedious task. I'm far over my quota on this. I'm stopping here.

I'll use the copy/paste version of Skype and use your template.

Now it seems the are calls missing in other platforms, my idea won't just work. Pretty bad :/

**Comment by Temptin** [comment]
This is the Messages table:

sqlite> .schema Messages
CREATE TABLE Messages (id INTEGER NOT NULL PRIMARY KEY, is_permanent INTEGER, convo_id INTEGER, chatname TEXT, author TEXT, from_dispname TEXT, author_was_live INTEGER, guid BLOB, dialog_partner TEXT, timestamp INTEGER, type INTEGER, sending_status INTEGER, consumption_status INTEGER, edited_by TEXT, edited_timestamp INTEGER, param_key INTEGER, param_value INTEGER, body_xml TEXT, identities TEXT, reason TEXT, leavereason INTEGER, participant_count INTEGER, error_code INTEGER, chatmsg_type INTEGER, chatmsg_status INTEGER, body_is_rawxml INTEGER, oldoptions INTEGER, newoptions INTEGER, newrole INTEGER, pk_id INTEGER, crc INTEGER, remote_id INTEGER, call_guid TEXT, extprop_chatmsg_ft_index_timestamp INTEGER, extprop_chatmsg_is_pending INTEGER, option_bits INTEGER, server_id INTEGER, annotation_version INTEGER, timestamp__ms INTEGER, language TEXT);

id - auto-incrementing ID
is_permanent - boolean value
convo_id - the conversation ID, ALWAYS exists
chatname - the title of the chat or something like that
author - who sent the message
from_dispname - the display name of that person at that time
author_was_live - boolean value
guid - used for mapping FILE TRANSFERS, not used for anything else. is a 32 bit binary blob; the Calls table has no corresponding BLOB field, so it isn't related to Calls at all.
dialog_partner - the skype id of the person you are talking to
timestamp - obvious
type - the event type
sending_status - delivered or not etc
consumption-status - read or not
edited_by - if the person edited it's their ID
edited_timestamp - when they edited the message
param_key - integer value, not sure what it is for, never touched it
param_value - integer again
body_xml - the message text
identities - list of people in the chat
reason - string, not sure what it's for
leavereason - integer, again not sure what it's for, seems new due to Microsoft
participant_count - how many people in a group chat
error_code - integer
chatmsg_type - old-school "event type", no longer used, is there for legacy purposes
chatmsg_status - old-school delivered/edited/etc, no longer used
body_is_rawxml - always 1, means the body text is XML code
oldoptions - integer, some legacy stuff
newoptions - integer, not sure
newrole - whatever... some permissions stuff...
pk_id - integer, some public key encryption id maybe... for skype's old p2p encryption
crc - not used, is a checksum
remote_id - something about a server or node ID
call_guid - a string describing the call ID...
extprop_chatmsg_ft_index_timestamp - integer timestamp
extprop_chatmsg_is_pending - integer pending status
option_bits - a bitmask with options
server_id - something with microsoft additions and delivery server i think
annotation_version - integer... whatever
timestamp__msg -- integer, whatever
language - text string, whatever

As you can see, there is exactly one way to find a call: Via call_guid.

Or the method I used before: Via Messages.convo_id == Calls.convo_dbid && Calls.begin_timestamp <= Messages.timestamp.

I am sorry that you and I got pulled into this. It was going to be a simple fix but Microsoft has just done something HORRIBLE here... and I don't see ANY way to find messages except what I've done...

I thought of possibly checking if the "call end" event timestamp is close to the begin_timestamp in the Calls database (the Calls database has no end timestamp unfortunately), and ONLY accepting recent call events so that it won't find/repeat old events... but then I realized that calls can have started hours before the "call end" event so that will not work either.

And this still doesn't solve how the hell to find calls now that they no longer log convo_dbid...

I wish I knew how the official client solved this mess, because they have done something evil.

**Comment by Temptin** [comment]
> Ok, I couldn't find the appropriate XKCD where a guy start working in a program so he avoid some tedious task. I'm far over my quota on this. I'm stopping here.

Me too. But just read the column descriptions in my message above, and you'll see how this makes no sense. I hope you can agree that there's no logic to this. The ONLY 3 pieces of info in the Messages table that could let us find calls are the call-end timestamp, the conversation ID, and the call GUID. Microsoft has dropped the conversation ID from the calls database, which is why we no longer find new calls. But they're seemingly not consistently writing GUIDs to the Calls table...

It's just a giant mess, and I promise this much:
This database is so broken now that I guarantee that the official Skype client has some weird mess of hacks that looks at multiple data fields to **guess** which call is the correct one. Probably trying GUID first, then if that fails it tries some timestamp + conversation partner guesswork.

I say fuck this and fuck programming.

Hey, @hectorpal... Now you know why I quit programming last year. It drives me mad. It's supposed to be a bit of quick fun and I end up in a rabbit hole. ;) Thanks for reminding me that I never want to program again in my life. :) At least not parsers for other people's buggy messes.

You're a nice guy and I wish you all the best. This is too much headache. None of this is your fault. It's Microsoft's fuckwit engineers. ;)

**Comment by Temptin** [comment]
https://xkcd.com/1319/ ... Yeah...

**Comment by hectorpal** [comment]
I started asking questions :) At least we know it's not easy to repair.

**Comment by hectorpal** [comment]
Yeap, that's the one...

**Comment by Temptin** [comment]
Yeah, a repair would ideally be done by looking at Skype's source code to see what they're doing, because it's definitely not anything logical. They're almost 100% guaranteed finding calls via some magic combo of "try via GUID. fails? try via conversation partner + timestamp. fails? try via conversation ID + timestamp, etc..." - and something like that is more than I have energy for. I am not paid enough for this... ;)

I will undo the damage to the repo and only keep your build script change, heh. Thanks for the idea. Take care and have a great summer. This is the last time I ever touch SkypeExport. Everything works except  calls. I can't be bothered wasting more hours of my life trying to figure out whatever patchy, hacked-together guess-method Microsoft has decided on using. It's nothing straight-forward like conversation ID or call ID anymore. They're... stupid.

If any archaeologists find this ticket in the future, the commits above have been deleted from the main repo, but they can still be accessed via their direct clickable links above, if anyone wants to analyze the two attempts at solving Microsoft's database bug. Then you know what should logically work but _won't_ work. The final solution, if someone _ever_ bothers spending the _days_ necessary to analyze the _deep_, messy corruption in the Skype database, will be something like what I described above: Try to find a GUID match, and if that fails, try to find some type of timestamp + conversation match, using a complex algorithm that needs to take into account what table columns are the most reliable for finding a correlation between the Calls table and Messages table... That reliability _used to_ be the Call GUID (Globally Unique Call ID), until Microsoft fucked up. And that's all folks... good luck, world. :)

Let these be my last words on this project: Microsoft are complete idiots. Ever since they took over Skype, they've constantly corrupted the database and broken things. If someone wants to maintain this project in the future, _don't_ be surprised if you run into more braindead Microsoft engineer bugs. Hopefully this call mismatch is their last database corruption for a long time, since everything else works... Good luck, world. I am never going to touch programming again. :)

**Comment by hectorpal** [comment]
I remember having lunch with master students of other disciplines. We started to talked about what our problems could be. There was a theoretical physicist saying removing a parenthesis of an equation took her a week, and then it may a lost of time. A chemistry saying he has problems in his experiments, and then found out the reactives were contaminated. The biology one makes us laugh: she was working on immunology and had been inoculating mice for three months, for the 1-year thesis. One day a cat got into the lab and eat them all. 

I guess every job has its details, and we go on where we can resist. The problem is the phase-transition: where it seems you can, and at the same time you are fed up about it. Those are the hard ones.

**Comment by Temptin** [comment]
Hahah thanks @hectorpal, the mouse story made me laugh out loud. That was a perfect end to these 3 hours of hair-pulling madness. Programming is okay when I just need to make a quick tool for myself, but when reverse engineering something that's as buggy as Skype, which is being rewritten with massive bugs by someone who never understood the codebase (Microsoft), it stops being fun and starts being insanity.

Skype's database has always been messy, with lots of legacy fields and weirdnesses. It was a patchwork of engineers trying to extend Skype's protocol for 7 versions, without breaking the old database format, so they kept adding "hacks" to make fields do new things, etc... But it was consistent, because they understood their own code. But Microsoft took the messiness to a whole new level when they bought Skype and wanted to "Microsoftify" it, without understanding the code they had bought. Their desktop client is pretty consistent in writing values to all correct columns, but their cloud history sync will NULL very important, vital columns in random cases (due to Microsoft only sending some of the db columns via the cloud, not all of them). It was hell to make it all work last summer. And this, today, takes the prize as the _worst_ bug they've ever introduced. I am done, never want to see Skype's database ever again. Hopefully Microsoft has stopped ruining it and won't ruin more things when they badly try to hack-in future Skype features. :)

But seriously, thanks for reminding me of why I quit programming. I just drifted onto GitHub today for some random reason, figured I'd log in, and saw the notification about this ticket, and was taken on this wild ride which made me want to puke. It's good to remember why I left this and went out into the sun instead. :) Much healthier.

I wish you a great summer, Hector!

**Comment by Temptin** [comment]
And to everyone else: I'm now throwing away my login password to this account, and setting the email to something random, so I won't ever be able to log in or recover the account again. Just so you know, this is it, I'm vanishing now. This is just so that I won't randomly be on GitHub for something else and think "may as well check what's going on with the old projects...", and find yet another seemingly "easy" tweak that's just going to be a massive nest of Microsoft bugs.

Take care all! Good luck with your lives and remember to live and love to the fullest! :)

**Comment by hectorpal** [comment]
good luck, @Temptin !

**Comment by hectorpal** [comment]
btw, the inconsistency between Desktop and mobile versions is deep. They behave almost as separated databases.
